File: 044_KeysAndRooms.py
"""
Topic: Graph / DFS
Problem: Keys and Rooms
Description: Determine if all rooms are reachable from room 0 given keys.
Time: O(n + sum(keys))
Space: O(n)
"""

from typing import List


def canVisitAllRooms(rooms: List[List[int]]) -> bool:
    n = len(rooms)
    seen = [False]*n
    stack = [0]
    seen[0] = True
    while stack:
        node = stack.pop()
        for k in rooms[node]:
            if not seen[k]:
                seen[k] = True
                stack.append(k)
    return all(seen)


if __name__ == '__main__':
    print(canVisitAllRooms([[1],[2],[3],[]]))  # True
    print(canVisitAllRooms([[1,3],[3,0,1],[2],[0]]))  # False

File: 044_KeysAndRooms.cpp
// Topic: Graph / DFS
// Keys and Rooms: Determine if all rooms can be visited starting from room 0.
// Time Complexity: O(n + e)
// Space Complexity: O(n)

#include <bits/stdc++.h>
using namespace std;

bool canVisitAllRooms(vector<vector<int>>& rooms)
{
    int n = rooms.size();
    vector<char> vis(n, 0);
    stack<int> st;
    st.push(0);
    vis[0] = 1;
    int cnt = 1;
    while (!st.empty())
    {
        int r = st.top();
        st.pop();
        for (int k : rooms[r])
            if (!vis[k])
            {
                vis[k] = 1;
                st.push(k);
                ++cnt;
            }
    }
    return cnt == n;
}

int main()
{
    vector<vector<int>> r1 = {{1}, {2}, {3}, {}};
    cout << (canVisitAllRooms(r1) ? "true" : "false") << '\n';  // true
    vector<vector<int>> r2 = {{1, 3}, {3, 0, 1}, {2}, {0}};
    cout << (canVisitAllRooms(r2) ? "true" : "false") << '\n';  // false
    return 0;
}
